### PR TITLE
upgrade postgres

### PR DESCRIPTION
**Pull request by ruslandoga**
This PR upgrades postgres to `postgres:14-alpine` for new setups.

---

For upgrade notes we can point users to existing blog posts / docs on upgrading postgres, [here's one using docker and upgrading from v12 to v14.](https://hollo.me/devops/upgrade-postgresql-database-with-docker.html)

This is what I did on my instance, I used `pg_dump + psql`:

1. Add a backup volume to `plausible_db` in your `docker-compose.yml`
```diff
  plausible_db:
    image: postgres:12
    restart: always
    volumes:
      - db-data:/var/lib/postgresql/data
+     - ./backup:/backup
    environment:
      - POSTGRES_PASSWORD=postgres
```

2. Stop all containers to avoid writing to old db

```
sudo docker compose down
```

3. Restart `plausible_db` container

```
sudo docker compose up plausible_db -d
```

4. Dump current DB contents to a backup file

```
sudo docker compose exec plausible_db pg_dump -U postgres plausible_db > /backup/plausible_db.bak
```

5. Edit `docker-compose.yml` to use new postgres version

```diff
  plausible_db:
-   image: postgres:12
+   image: postgres:14-alpine
    restart: always
    volumes:
      - db-data:/var/lib/postgresql/data
      - ./backup:/backup
    environment:
      - POSTGRES_PASSWORD=postgres
```

6. Stop `plausible_db`

```
sudo docker compose down
```

7. Remove old data volume

```console
> sudo docker volume ls
DRIVER    VOLUME NAME
local     0f3bfd36a9b5e0bd2202469d57957e4a5a8d9b6623f7d8c23a14fc94c2afa425
local     5552ed5363bb7097ff0d68796f55ac38270ac414752fcf1a3d7238961a2420e6
local     a60745b6fbb320f95b4da7f281bb1ae1905be9bd19808bb17bf55a567ec09497
local     ce7ec303fa9ed9f73a660fb9df15f3b296aed9f64167bcf09c1cadf1a335f7ac
local     hosting_db-data
local     hosting_event-data

> sudo docker volume rm hosting_db-data
hosting_db-data
```

8. Start new db

```
sudo docker compose up plausible_db -d
```

9. Create new db and load data into it

```
sudo docker compose exec plausible_db createdb -U postgres plausible_db
sudo docker compose exec plausible_db psql -U postgres -d plausible_db < /backup/plausible_db.bak
```

10. Start all other containers

```
sudo docker compose up -d
```

11. (Optional) Remove backups from `docker-compose.yml` and your filesystem

```diff
  plausible_db:
    image: postgres:14-alpine
    restart: always
    volumes:
      - db-data:/var/lib/postgresql/data
-     - ./backup:/backup
    environment:
      - POSTGRES_PASSWORD=postgres
```

```
> rm -rf ./backup
```

### CODE REVIEW

**Comment by paskal** [comment]
That should be quoted:

```
$ docker compose exec plausible_db pg_dump -U postgres plausible_db > /backup/plausible_db.bak
-bash: /backup/plausible_db.bak: No such file or directory
```
I think it should be `docker compose exec plausible_db sh -c "pg_dump -U postgres plausible_db > /backup/plausible_db.bak"` instead. Same goes for `docker compose exec plausible_db sh -c "psql -U postgres -d plausible_db < /backup/plausible_db.bak"`

Also this doesn't work unless you remove old container:

```
$ docker volume rm hosting_db-data
Error response from daemon: remove hosting_db: volume is in use - [6168247f613eb9eaf25886eecdc0995a8264318f289690c7940142177dad6652]
```

`docker rm` by id would work, or better by the name if that's possible to make the instruction more generic.

Also I wouldn't recommend putting commands like `sudo docker compose down` into the documentation, `docker compose stop plausible plausible_db plausible_events_db` and `docker compose stop plausible_db` should be used instead.

I assume it's okay to bump the PostgreSQL version here as it won't be propagated to users unless they change their own configurations. However, that instruction should preferably be posted in Readme.

**Comment by ruslandoga** [comment]
@paskal thank you for fixing my mistakes!

I've added a guide at `upgrade/postgres.md` (with your fixes 🥰) and linked to it from README.md.

**Comment by ruslandoga** [comment]
@ukutaht 👋

What do you think?

File: README.md [hunk 1]
```diff
@@ -10,3 +10,7 @@ Find instructions on how to run Plausible Analytics Self Hosted [in our docs](ht
 
 We are always looking to expand on the options and setups provided here. Feel free to open an issue or PR if you feel
 something could be improved.
+
+### Upgrade guides
+
+- [Upgrading `plausible_db` (PostgreSQL)](upgrade/postgres.md)
```
Comment (paskal) [inline README.md:16]: I would propose writing "from version 12 to version 14" explicitly there, as that's precisely why we need this documentation in the first place.

**Review by paskal — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (ruslandoga) [reply README.md:16]: The idea here is that this guide can be used to upgrade from any version, hence the `pg_dump + psql` choice. In the release notes we can specify that we recommend existing installations to be updated from 12 to 14.

Comment (ukutaht) [reply README.md:16]: The upgrade guide should work with any PostgreSQL version >12.

We can document the supported PostgreSQL versions (12,13,14). I've added all 3 into our unit testing matrix as well: https://github.com/plausible/analytics/commit/eca71b9b60db288bfba1f366ceb4bccf8b0a4f98

Comment (ukutaht) [reply README.md:16]: So I think to address the original concern, I would propose adding something like:

`Supported PostgreSQL versions are 12,13, and 14`

File: README.md [hunk 2]
```diff
@@ -10,3 +10,9 @@ Find instructions on how to run Plausible Analytics Self Hosted [in our docs](ht
...
+
+  Supported versions are 12, 13, and 14
```
Comment (ruslandoga) [inline README.md:18]: I've added a versions disclaimer to the readme here.

File: upgrade/postgres.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
+Guide to upgrading `postgres` from versions 12 or 13 to version 14 using `pg_dump` and `psql`. Based on [Upgrade a PostgreSQL database with docker.](https://hollo.me/devops/upgrade-postgresql-database-with-docker.html)
```
Comment (ruslandoga) [inline upgrade/postgres.md:1]: And I've added a versions disclaimer to the guide here.

File: upgrade/postgres.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
+Guide to upgrading PostgreSQL version `>12` to `14` using `pg_dump` and `psql`. Based on [Upgrade a PostgreSQL database with docker.](https://hollo.me/devops/upgrade-postgresql-database-with-docker.html)
```
Comment (ruslandoga) [inline upgrade/postgres.md:1]: I've added versions info here.

File: docker-compose.yml [hunk 5]
```diff
@@ -5,7 +5,8 @@ services:
     restart: always
 
   plausible_db:
-    image: postgres:12
+    # supported versions are 12, 13, and 14
```
Comment (ruslandoga) [inline docker-compose.yml:8]: And here. Not sure if it needs to be added anywhere else?

File: README.md (hunk 2 above)
Comment (ruslandoga) [reply README.md]: I moved it to `docker-compose.yml`

**Comment by ruslandoga** [comment]
@paskal what do you think, does it make it a bit more clear to the user?

File: upgrade/postgres.md (hunk 4 above)
Comment (ukutaht) [reply upgrade/postgres.md:1]: > Guide to upgrading PostgreSQL version `>12` to `14`

Hmm this is a bit confusing if I'm currently running 12. Shouldn't it be `version >= 12 to 14`?

Comment (ruslandoga) [reply upgrade/postgres.md]: Fixed.

**Review by paskal — APPROVED** [review]

File: docker-compose.yml (hunk 5 above)
Comment (ruslandoga) [reply docker-compose.yml:8]: Hm, but it's already `latest`.

**Pull request merged**